pepper-apis/src/main/java/org/broadinstitute/ddp/service/ActivityInstanceService.java

```diff
@@ -551,8 +566,71 @@ public void deleteInstance(Handle handle, ActivityInstanceDto instanceDto) {
             String instanceGuid,
             ContentStyle style,
             String isoLangCode) {
-        return new ActivityInstanceFromDefinitionBuilder().buildActivityInstance(
-                handle, userGuid, operatorGuid, studyGuid, instanceGuid, style, isoLangCode
-        );
+
+        var context = new ActivityInstanceFromDefinitionBuilder(handle,
+                createParams(userGuid, studyGuid, instanceGuid)
+                        .setOperatorGuid(operatorGuid)
+                        .setIsoLangCode(isoLangCode)
+                        .setStyle(style))
+                .checkParams()
+                    .readFormInstanceData()
+                    .readActivityDef()
+                .startBuild()
+                    .buildFormInstance()
+                    .buildFormChildren()
+                    .renderFormTitles()
+                    .renderContent()
+                    .setDisplayNumbers()
+                    .updateBlockStatuses()
+                .endBuild()
+                    .getContext();
+
+        if (context.getFailedStep() != null) {
+            LOG.warn("ActivityInstance build failed: {}, step={}", context.getFailedMessage(), context.getFailedStep());
+        }
+        return Optional.ofNullable(context.getFormInstance());
+    }
+
+    /**
+     * Build {@link ActivityInstance} from data cached stored in {@link ActivityDefStore}.
+     * Some of data (answers, rule messages) are queried from DB.
+     * This method provide partial building and rendering of {@link FormInstance}, plus it passes 'instanceSummary'.<br>
+     * The following building steps executed:
+     * <pre>
+     * - create form;
+     * - add children;
+     * - render form title/subtitle;
+     * - update block statuses.
+     * </pre>
+     */
+    public Optional<ActivityInstance> buildInstanceFromDefinition(
+            Handle handle,
+            String userGuid,
+            String operatorGuid,
+            String studyGuid,
+            String instanceGuid,
+            String isoLangCode,
+            UserActivityInstanceSummary instanceSummary) {
+
```

Comment:
This looks nice, but I don't think this buys us much. What we wanted in the PUT route was to avoid rendering templates. We only need to load the form structure, load the answer data, and the block statuses -- that's all we need to know to figure out whether the user has completed the form or not; no need to render the templates since we're not displaying anything.

However, the current implementation of AIBuilder is always rendering templates, even with this "builder pattern" refactor. That's because when building the "form instance" and the "form children", every object we create as part of the structure will call `addAndRenderTemplate()` which will basically render each template we encounter during traversal. It's not exactly "rendering the form" since they are stashed away in the context and is not actually put into the form structure objects. But still, that's work we don't want to pay for.

Anyways, what I'm trying to say here is, this refactor doesn't actually help us much. I think if we skipped this "builder pattern" refactor and just used AIBuilder as-is in the PUT route, the performance will probably be very similar to what this refactor gives us.

But I could be wrong! Maybe I've misunderstood, and that this is part of a bigger upcoming refactor? Regardless, this "builder pattern" style usage of AIBuilder might be useful in other cases, so I'll let you decide how you want to proceed here.

Reply:
Ok, I understood the idea with templates. I will add a parameter to turn off templates rendering.
Another place where this builder could be used (or at least a possibility to call building steps separately) - in GetActivityInstanceRoute: to replace `RouteUtil.findAccessibleInstanceOrHalt()` with the 1st step of AIBuilder: readFormInstanceData() which reads all data which reads `RouteUtil.findAccessibleInstanceOrHalt()` (or actually reads `.getByActivityInstanceGuid(instanceGuid)` which called from `RouteUtil.findAccessibleInstanceOrHalt()` ).
So we can avoid yet one extra query in `GetActivityInstanceRoute`.

Reply:
Added the flag to disable templates rendering and used it for `PutFormAnswerRoute`.

Reply:
Looks good!